src/Stratis.Bitcoin/P2P/Peer/NetworkPeerListener.cs

```diff
@@ -40,15 +40,18 @@ public NetworkPeerListener Where(Func<IncomingMessage, bool> predicate)
             Queue<IncomingMessage> pushedAside = new Queue<IncomingMessage>();
             try
             {
-                while (true)
+                using (CancellationTokenSource cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, this.Peer.Connection.Cancel.Token))
                 {
-                    IncomingMessage message = ReceiveMessage(CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, this.Peer.Connection.Cancel.Token).Token);
-                    if (this.predicates.All(p => p(message)))
+                    while (true)
```

Comment:
perhaps?
``` 
while (true)
{
       cancellation.ThrowIfCanceled() ;
```

Reply:
this is unchanged code and I think it just relies on ReceiveMessage throwing
and that's probably correct assumption, RM here would block, so it is its responsibility, such cancellation.throw  would not help us anything i believe, but this will be rewritten later anyway, but should work perfectly as is